src/metrics/chaoss.ts

```diff
@@ -539,3 +540,95 @@ FORMAT JSONCompact`;
     }
   });
 };
+
+  interface NewContributorsOptions {
+    // calculate new contributors by change request or git commit index. default: change request
+    by: 'commit' | 'change request';
+    withBot: boolean;
+  }
+  export const chaossNewContributors = async (config: QueryConfig<NewContributorsOptions>) => {
+    config = getMergedConfig(config);
+    const by = filterEnumType(config.options?.by, ['commit', 'change request'], 'change request');
+    const whereClauses: string[] = [];
+    const endDate = new Date(`${config.endYear}-${config.endMonth}-1`);
+    endDate.setMonth(config.endMonth);  // find next month
+    if (by === 'commit') {
+      whereClauses.push("type = 'PushEvent'")
+    } else if (by === 'change request' ) {
+      whereClauses.push("type = 'PullRequestEvent' AND action = 'closed' AND pull_merged = 1");
+    } 
+    const repoWhereClause = getRepoWhereClauseForClickhouse(config);
+    if (repoWhereClause) whereClauses.push(repoWhereClause);
+    const sql = `
+  SELECT
+    id,
+    argMax(name, time) AS name,
+    ${getGroupArrayInsertAtClauseForClickhouse(config, { key: 'new_contributors', })},
+    ${getGroupArrayInsertAtClauseForClickhouse(config, { key: 'detail' })}
+  FROM
+  (
+    SELECT
+    ${getGroupTimeAndIdClauseForClickhouse(config, 'repo','first_time')},
+      length(detail) AS new_contributors,
+      (arrayMap((x) -> (x), groupArray(author))) AS detail
+    FROM
+    (
+      SELECT
+        min(created_at) AS first_time,
+        repo_id,
+        argMax(repo_name,created_at) AS repo_name,
+        ${(() => {
+          if (by === 'commit') {
+            return `
+            author
+            `
+          } else if (by === 'change request') {
+            return `
+            actor_id,
+            argMax(author,created_at) AS author
+            `
+          } 
+          })()}
+      FROM
+       (
+          SELECT 
+            repo_id,
+            repo_name,
+            ${(() => {
+            if (by === 'commit') {
+              return `
+              arrayJoin(push_commits.name) AS author
+              `
+            } else if (by === 'change request') {
+              return `
+              issue_author_id AS actor_id,
+              issue_author_login AS author
+              `
+            } 
+            })()},
+            created_at
+          FROM github_log.events
+          WHERE ${whereClauses.join(' AND ')}
+          ${(config.options?.withBot && by !== 'commit') ? '' : "HAVING author NOT LIKE '%[bot]'"}
+        )
+      GROUP BY repo_id, ${by === 'commit' ? 'author' : 'actor_id' }
+      HAVING first_time >= toDate('${config.startYear}-${config.startMonth}-1') AND first_time < toDate('${endDate.getFullYear()}-${endDate.getMonth() + 1}-1')
+    )
+    GROUP BY id, time
```

Comment:
If `limit` option is set, you should also add a `LIMIT` clause here to avoid more data returned. I think I got `bus_factor` implemented wrong, actually we don't need a `LIMIT` clause outside but only here.

Or maybe we should discuss the logic of how should we implement limit option.

Reply:
I think add `limit` outside and add it here means two different implementation logics.

For example,if we use `labelIntersect: ['Company'], limit: 10` ,the first one is to find 10 repos with the most new contributors among all repos, and the second one is to randomly select 10 repos and sort them in descending order.

Is my understanding correct？Which logic do you think is better？

Reply:
Not really.

`LIMIT` outside means calculate the metric for all repos and all time period and return top results ordered by the last value.

`LIMIT` inside should also contained with `ORDER` clause which means find top results in each time period and return  the list ordered by the last value.

The first one is better because it will calculate all the value but it is really resource consuming, the second is more economic but some value maybe lost.

Reply:
The first one returns all the data ordered by last value, so any top-10 item may not be included in the result since it is not one of the top 10 items in the last time period.

The second one will return all top 10 items for all time period which makes sense for applications like OpenLeaderboard.

Reply:
If we `limit` outside, we can order by their total new contributors and limit. This can avoid some situations like a repo has a lot contributors before but none contributors in the last period. We can implement this quickly.
![image](https://user-images.githubusercontent.com/91717658/203279454-6abbec8c-f2a2-42a5-98c6-e1d421972b44.png)

If `limit` inside, what we do are `In a specific time period, which companies have the most new contributors. Is this right?
We will try to implement this tonight.

Reply:
Yes, I think we should use outside limit, that makes sense a lot.